### PR TITLE
Generalize Hut Checks

### PR DESCRIPTION
**Pull request by Justen-N**
# Changes proposed in this pull request:
- Implements a more generalized check for hut blocks 
- Instead of limiting check to the Minecolonies namespace, the check is now performed against each origin mod's namespace. 
-

Review please

### CODE REVIEW

File: src/main/java/com/ldtteam/structurize/management/StructureName.java [hunk 1]
```diff
@@ -106,13 +109,14 @@ else if (name.startsWith(Structures.SCHEMATICS_CACHE + Structures.SCHEMATICS_SEP
...
             section = Structures.SCHEMATICS_PREFIX;
-            if (ForgeRegistries.BLOCKS.getValue(new ResourceLocation(Constants.MINECOLONIES_MOD_ID, "blockhut" + hut)) != Blocks.AIR || HUTS.contains(hut))
-            {
-                section = hut;
-            }
-            else
-            {
-                hut = "";
+            //This allows mods other than minecolonies to create hut buildings and use the buildtool with them.
+            for (final Map.Entry<String, ModFileInfo> origin : StructureLoadingUtils.getOriginMods().entrySet()) {
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/management/StructureName.java]: Seems okay, just the formatting is totally off.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: What about the formatting needs fixed? Other than my somewhat unnecessary brackets? (former C developer, I'm a bracket addict.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: I think I see it now. I'm dumb

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: Wonder where the setting in Intellij that is f-ing with that is...

File: src/main/java/com/ldtteam/structurize/management/StructureName.java [hunk 2]
```diff
@@ -106,13 +109,18 @@ else if (name.startsWith(Structures.SCHEMATICS_CACHE + Structures.SCHEMATICS_SEP
         {
             hut = schematic.split("\\d+")[0];
             section = Structures.SCHEMATICS_PREFIX;
-            if (ForgeRegistries.BLOCKS.getValue(new ResourceLocation(Constants.MINECOLONIES_MOD_ID, "blockhut" + hut)) != Blocks.AIR || HUTS.contains(hut))
+            //This allows mods other than minecolonies to create hut buildings and use the buildtool with them.
+            for (final Map.Entry<String, ModFileInfo> origin : StructureLoadingUtils.getOriginMods().entrySet())
```
Comment (Nightenom) [inline src/main/java/com/ldtteam/structurize/management/StructureName.java]: You should be fine with key view (you don't seem to use value from entry set)

**Review by Nightenom — DISMISSED** [review]

File: src/main/java/com/ldtteam/structurize/management/StructureName.java (hunk 2 above)
Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: You are absolutely right.

**Review by Raycoms — DISMISSED** [review]

File: src/main/java/com/ldtteam/structurize/management/StructureName.java [hunk 3]
```diff
@@ -106,13 +109,18 @@ else if (name.startsWith(Structures.SCHEMATICS_CACHE + Structures.SCHEMATICS_SEP
...
-                hut = "";
+                if (ForgeRegistries.BLOCKS.getValue(new ResourceLocation(origin, "blockhut" + hut)) != Blocks.AIR || HUTS.contains(hut))
+                {
+                    section = hut;
+                    break;
+                }
+                else
+                {
+                    hut = "";
+                }
```
Comment (marchermans) [inline src/main/java/com/ldtteam/structurize/management/StructureName.java]: If this is meant to work for other mods, then we should change this mechanism.
Use some form of marker interface instead of using a `blockhut` prefix. or a map that is named `HUT` since this is then really not clear.

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/com/ldtteam/structurize/management/StructureName.java (hunk 3 above)
Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: @ShadowProtocol is doing a bigger rework with propper metadata files which will contain the propper descriptions for the future. But that rework will take a while longer. This is meant to work for the time being as a replacement for @Justen-N addon to work propperly.

Comment (marchermans) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: That has nothing todo with a rework.
This just needs to be defined better.

So no. Make it a marker interface on the block and call it a day.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: This was a 'quick fix' (read: bad idea) for a straight-across companion mod ( herder compatibility) But I can see where you are coming from.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: I guess, my question regarding a marker interface, is where would we put it? In Structurize? It feels like that would force an even tighter coupling between Minecolonies and Structurize.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: I suppose we could check if the block extends the AbstractBlockHut class, if that works for you.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: By the way, at the moment, all of minecolonies hut blocks still require the blockhut prefix, which is why I did not change that portion of the check. If we remove it, we'll have to refactor the blocks from Minecolonies.

File: src/main/java/com/ldtteam/structurize/management/StructureName.java [hunk 4]
```diff
@@ -106,13 +108,18 @@ else if (name.startsWith(Structures.SCHEMATICS_CACHE + Structures.SCHEMATICS_SEP
...
-            if (ForgeRegistries.BLOCKS.getValue(new ResourceLocation(Constants.MINECOLONIES_MOD_ID, "blockhut" + hut)) != Blocks.AIR || HUTS.contains(hut))
-            {
-                section = hut;
-            }
-            else
+            //This allows mods other than minecolonies to create hut buildings and use the buildtool with them.
+            for (String origin : StructureLoadingUtils.getOriginMods().keySet())
             {
-                hut = "";
+                if (IAnchorBlock.class.isAssignableFrom(ForgeRegistries.BLOCKS.getValue(new ResourceLocation(origin, "blockhut" + hut)).getClass()))
```
Comment (Nightenom) [inline src/main/java/com/ldtteam/structurize/management/StructureName.java]: IAnchorBlock is definitely not a way for me to distinguish huts and the others
@OrionDevelopment meh, reworking this to something "normal" does not make much sense - the mod is one gigantic tape fix and every time you try to rewrite something you end up rewriting half of the mod, I'm completely fine with previous minor change

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: I'm willing to revert this. My reasoning was that every hut block implements IAnchorBlock. I do realize that decoration controllers do so as well, but figured that they would be placed elsewhere by StructureLoadingUtils because of the path used .

**Comment by Justen-N** [comment]
@OrionDevelopment, I'm leaving this in the state you originally reviewed it in. Changing this to be a 'true' implementation is a loss of effort that could be put in elsewhere, especially because this check will need to be changed completely alongside the metadata changes that are planned. For the moment, the simple solution proposed here is, in effect, correcting for technical debt in the short term, until that debt can be fully addressed later on.

File: src/main/java/com/ldtteam/structurize/management/StructureName.java [hunk 5]
```diff
@@ -106,13 +109,18 @@ else if (name.startsWith(Structures.SCHEMATICS_CACHE + Structures.SCHEMATICS_SEP
...
-            }
-            else
-            {
-                hut = "";
+                if (ForgeRegistries.BLOCKS.getValue(new ResourceLocation(origin, "blockhut" + hut)) != Blocks.AIR || HUTS.contains(hut))
+                {
+                    section = hut;
+                    break;
+                }
+                else
```
Comment (Nightenom) [inline src/main/java/com/ldtteam/structurize/management/StructureName.java]: This is not a correct logic rewrite btw, if you iterate first then you end up with hut = ""

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: ... Did I seriously.... damn.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: Alright, I just removed the else block. From what I can see, it isn't necessary anyway.

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: Even now it's not correct :D you are missing the else branch logic now

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: And also the list is probably needed? Depends on usage in mcol

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: .... Wow. I'm just gonna go sit in a corner now.

Comment (Justen-N) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java]: @Nightenom, the HUTS list is not referenced from minecolonies. I'm pretty sure it was originally populated so that it could be checked against by the BuildTool. I went through each time StructureName is imported by mcol, it's never used.

File: src/main/java/com/ldtteam/structurize/management/StructureName.java [hunk 6]
```diff
@@ -106,13 +108,18 @@ else if (name.startsWith(Structures.SCHEMATICS_CACHE + Structures.SCHEMATICS_SEP
         {
             hut = schematic.split("\\d+")[0];
             section = Structures.SCHEMATICS_PREFIX;
-            if (ForgeRegistries.BLOCKS.getValue(new ResourceLocation(Constants.MINECOLONIES_MOD_ID, "blockhut" + hut)) != Blocks.AIR || HUTS.contains(hut))
+            //This allows mods other than minecolonies to create hut buildings and use the buildtool with them.
+            for (String origin : StructureLoadingUtils.getOriginMods().keySet())
             {
-                section = hut;
+                if (ForgeRegistries.BLOCKS.getValue(new ResourceLocation(origin, "blockhut" + hut)) != Blocks.AIR)
```
Comment (marchermans) [inline src/main/java/com/ldtteam/structurize/management/StructureName.java:114]: Again this won't be merged in the current state.
I talked this through with Ray.

Since Structurize 2.0 is a long long long long long way away, we have to work with this.
And since a standing (partially unwritten) rule for this mod is that as long as it is for Minecolonies then everything is allowed. But since PR in particular is targeted at supporting other mods, this won't fly and the exception does not take here.

In other words, create a proper API for this instead of hacking around.

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/com/ldtteam/structurize/management/StructureName.java (hunk 6 above)
Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java:114]: I wrote in the other side. Let's remove all the hut stuff from this StructureName class, make in MineColonies an implementation that overrides this Structurename and handle it there.

Comment (marchermans) [reply src/main/java/com/ldtteam/structurize/management/StructureName.java:114]: Also possible.

**Comment by Justen-N** [comment]
I talked with ray as well, and agreed that we will handle this on the
minecolonies side. I'm out of town for a week, but will pick this thread up
later on.
